### ISSUE TITLE
Baum-Welch for multinomial HMM converges to local optima

### ISSUE THREAD

**Issue by kolchinski** [hmmlearn/hmmlearn#162]
In generating sequences from a multinomial (really, Bernoulli) HMM and then learning the parameters back with hmmlearn, the parameters are learned roughly correctly some of the time (about half) and very incorrectly the rest of the time. This holds true regardless of the number of sample sequences (numSeqs in the code below). For reference, Matlab's HMM fitting function correctly fits the parameters for the same model every time it is run. My best guess is that the Baum-Welch algorithm underlying the parameter fitting in hmmlearn is converging to local optima. It would be very helpful if there were safeguards to break out of local optima so that parameter fitting could be counted upon to work every time.

`import numpy as np

from hmmlearn.hmm import MultinomialHMM

pT = .1
pL0 = .3
pG = .25
pS = .05

model = MultinomialHMM(n_components=2)
model.startprob_  = np.array([1.0-pL0, pL0])
model.transmat_ = np.array([[1.0-pT,pT],[0,1]])
model.n_features = 2
model.emissionprob_ = np.array([[1.0-pG, pG], [pS, 1.0 - pS]])

numSeqs = 100

for n in range(10):
  print '\niteration ', n

  seqs = []
  seqLengths = []

  for i in range(numSeqs):
#seqLength = np.random.randint(10,50)
    seqLength = 20
    seqLengths.append(seqLength)
    seq = model.sample(seqLength)[0]
    if len(seqs) > 0:
      seqs = np.append(seqs, seq, axis=0)
    else: seqs = seq

  seqs = np.array(seqs)
  seqLengths = np.array(seqLengths)

  model2 = MultinomialHMM(n_components=2)
  model2.fit(seqs, seqLengths)
  print 'transition: '
  print model2.transmat_
  print 'emission: '
  print model2.emissionprob_

`

**Comment by superbobry** [comment]
Baum-Welch is indeed susceptible to local optima. 

Do you know how the Matlab version does initialization? Maybe that could explain the differences? Also, do you use the same dataset for Matlab runs?

**Comment by kolchinski** [comment]
Unfortunately I'm not sure - I believe Matlab is closed source but guessing the Octave version of the hmmtrain function (https://octave.sourceforge.io/statistics/index.html) should work the same way. My best guess is that their implementation tries many different start conditions and then of the final values arrived at post-convergence, picks the one that corresponds to the highest likelihood (i.e. picking the best local maximum). For every set of training data I've tried with Matlab, hmmtrain has arrived at the same fitted parameters every time it's executed, which has not been the case for the hmmlearn fitting. It seems that te matlab implementation is successfully finding global optima, or at least the same (very good) local optima every time. 
I'm very short on time right now (final exams) so I can't do much more debugging but I can have another look in a couple of weeks if that would be helpful.

**Comment by superbobry** [comment]
You're right, multi-start is a common trick when dealing with EM-algorithm. We could have a multi start helper function in `hmmlearn`. I'd welcome a PR.

**Comment by hkoohy** [comment]
I also have the same problem.
In the following toy example, I need to be able to get the original parameters that I used to generate the sequences in the first place, but I fail:
`
def initialize_components(n_components,  n_emissions):
    """
    Initialize start probabilities,  transition and emission matrices, start  for a model with a fixed number of components,
    for Multinomial model with a certain number of dimensions.
    """

    trans_mat = np.abs(np.random.randn(n_components, n_components))
    trans_mat = (trans_mat.T / trans_mat.sum( axis=1 )).T
    
    emission_mat = np.abs(np.random.randn(n_components, n_emissions))
    
    emission_mat = (emission_mat.T/emission_mat.sum(axis=1)).T
#     print(emission_mat.shape)

    start_probs = np.abs( np.random.randn(n_components) )
    start_probs /= start_probs.sum()
    return (start_probs, trans_mat, emission_mat)

def generate_sequences(Pi, T, E, n_components=3, n_features=6, N=200, K=20):
    """
    Use the given parameters, initiate a hmm model and generate some sequences
    """
    states = []
    seqs = []
    seq_lengths = []
    
    model = hmm.MultinomialHMM(n_components=3)
    model.startprob_ = Pi
    model.transmat_ = T
    model.emissionprob_ = E
    model.n_features = n_features
    print('*** Original parameters *********')
    print('p_init')
    print(model.startprob_)
    print('p_tans ')
    print(model.transmat_)
    print('p_emm')
    print(model.emissionprob_)

    
    for j in range(N):
        seq_length = K 
        ### generate sequence
        seq = model.sample(seq_length)[0]
        z = model.sample(seq_length)[1]
        #     print(z)
        # print(seq)
        ### generate sequence
        seq = model.sample(seq_length)[0]
        seq_lengths.append(seq_length)
        states.append(z)
        # print(seq)
        if len(seqs) > 0:
            seqs = np.append(seqs, seq, axis=0)
        
        else: 
            seqs = seq
            
    model.fit(seqs, seq_lengths)
    score =  model.score(seqs)
    print('Score: ', score)
    return(seqs, states, seq_lengths)

def train(hmm_model, Seqs, Lengths, nIteration):
    """
    Train the model with given data, and find the best describing parameters (Baub-Welch)
    """
    best_model = None
    best_trans = None
    best_emissions = None
    best_pi = None
    max_score = -np.inf

    for i in range(nIteration):
        
        pi, A, B = initialize_components(n_states, n_emissions)
        hmm_model.emissionprob_ = B
        hmm_model.transmat_ = A
        hmm_model.startprob_ = pi
        hmm_model.n_features_ = 6
        hmm_model.fit(Seqs, Lengths)
        this_score =  hmm_model.score(Seqs)

        if this_score > max_score:
            max_score = this_score
            best_model = hmm_model
            best_pi = pi
            best_trans = A
            best_emissions = B
         ### used these start poing
        if i % 10 == 0:
            print('^'*50)
            print('\niteration', i)
            print('best score: ', max_score)


#     print(i , '\n ================== After training ==============')
#     print(model2.score(seqs))

    print(max_score)
    print(best_model.startprob_)
    print(best_model.transmat_)
    print(best_model.emissionprob_)
    return(best_model)

S = ["S1", "S2", 'S3']
n_states = len(S)
E = [i+1 for i in range(6)] ## sides of a dice
trans_prob = np.array(
    [[0.9,  0.05, 0.05], 
     [0.1, 0.8, 0.1], 
     [0.05,  0.05, 0.9]]
)
start_prob = np.array([0.8, 0.1, 0.1])

### Lets assume that we are throwing 3 6-sided dices: one is fair with equel probablity of each side, and the other two are loaded
em_prob =  np.array([[1/4, 3/20, 3/20, 3/20, 3/20, 3/20],
                [2/15, 2/15, 2/15, 2/15, 2/15, 1/3],
                [1.0/6, 1.0/6, 1.0/6, 1.0/6, 1.0/6, 1.0/6]])

X, Z, L = generate_sequences(start_prob, trans_prob, em_prob, n_components=3, n_features=6, N=200, K=20)
model2 = hmm.MultinomialHMM(n_components=3,   n_iter=200)

fitted_hmm = train(model2, X, L, 10)
`
